components/engine_traits/src/engine.rs

```diff
@@ -63,5 +66,101 @@ pub trait KvEngine:
 // It should be named as `EngineFactory` for consistency, but we are about to rename
 // engine to tablet, so always use tablet for new traits/types.
 pub trait TabletFactory<EK> {
-    fn create_tablet(&self) -> Result<EK>;
+    fn loop_tablet_cache(&self, _f: Box<dyn FnMut(u64, u64, &EK) + '_>) {}
+    fn destroy_tablet(&self, _id: u64, _suffix: u64) -> crate::Result<()> {
+        Ok(())
+    }
+    fn create_tablet(&self, id: u64, suffix: u64) -> Result<EK>;
+    fn open_tablet(&self, id: u64, suffix: u64) -> Result<EK> {
+        self.open_tablet_raw(&self.tablet_path(id, suffix), false)
+    }
+    fn open_tablet_cache(&self, id: u64, suffix: u64) -> Option<EK> {
+        if let Ok(engine) = self.open_tablet_raw(&self.tablet_path(id, suffix), false) {
+            return Some(engine);
+        }
+        None
+    }
+    fn open_tablet_cache_any(&self, id: u64) -> Option<EK> {
+        self.open_tablet_cache(id, 0)
+    }
+    fn open_tablet_raw(&self, path: &Path, readonly: bool) -> Result<EK>;
+    fn create_shared_db(&self) -> Result<EK>;
+    #[inline]
+    fn exists(&self, id: u64, suffix: u64) -> bool {
+        self.exists_raw(&self.tablet_path(id, suffix))
+    }
+    fn exists_raw(&self, path: &Path) -> bool;
+    fn tablet_path(&self, id: u64, suffix: u64) -> PathBuf;
+    fn tablets_path(&self) -> PathBuf;
+    fn clone(&self) -> Box<dyn TabletFactory<EK> + Send>;
+    fn load_tablet(&self, _path: &Path, id: u64, suffix: u64) -> Result<EK> {
+        self.open_tablet(id, suffix)
+    }
+    fn mark_tombstone(&self, _region_id: u64, _suffix: u64) {}
+    fn is_tombstoned(&self, _region_id: u64, _suffix: u64) -> bool {
+        false
+    }
+}
```

Comment:
I think we should introduce new method only when it's truly used.

Reply:
per the prototype we did, these methods should be used.
The tombstone is used for releasing the tablet.

Reply:
I mean adding it in the PR that uses it. I think it makes the context clearer and won't add dev overhead.

Reply:
So the approach is we define the engine function first and then will implement other components. These functions can be used by different PRs in the same time. 
Because we're implementing a new version of raftstore and so it will be more efficient to define the APIs first. Otherwise, there's no need for this PR at all.

Reply:
In that case I presume you want it to be close the final version? How about use doc comments and order the methods properly (and consistently).

Reply:
fair.